src/fft.rs

```diff
@@ -58,15 +58,26 @@ pub fn discrete_fourier_transform<F: FftFriendlyFieldElement>(
         w = F::one();
         let r = F::root(l).unwrap();
         let y = 1 << (l - 1);
-        for i in 0..y {
-            for j in 0..(size / y) >> 1 {
-                let x = (1 << l) * j + i;
+        let chunk = (size / y) >> 1;
+
+        // unrolling first iteration of i-loop.
+        for j in 0..chunk {
+            let x = j << l;
+            let u = outp[x];
+            let v = outp[x + y];
+            outp[x] = u + v;
+            outp[x + y] = u - v;
+        }
+
+        for i in 1..y {
+            w *= r;
+            for j in 0..chunk {
+                let x = (j << l) + i;
```

Comment:
It looks to me like `x` is computed differently. Previously we had $x = 2^\ell \cdot j + i$; now it is $x = j^2 + i$. It's been a while since i had FFT in my head, but why does the new math work?

Reply:
It's equivalent, note that the shift amount is the letter l, not one. The one change this line makes is replacing multiplication by the result of left shifting one with just a left shift.

Reply:
Ohohh, misread `l` as `1` 😬